### PR TITLE
Servlet logging improvements

### PR DESCRIPTION
**Pull request by rvanheest**
These are the first improvements that resulted from the servlet-logging workshop of last week. See the individual commits for what changed.

* merge auth headers and action headers in the response log
* request/response prefix for every log statement
* LogResponseBody functionality (see [example](https://github.com/DANS-KNAW/dans-scala-lib/blob/12f40a0593ed89a89656a5930184c54b6d81ee29/src/test/scala/nl/knaw/dans/lib/logging/servlet/examples/LogWithBodyOnErrorServlet.scala))
* remove `MaskedAuthenticationParameters`

The other points will be done later, as described in EASY-1992 and EASY-1993.

@DANS-KNAW/easy for review

* [x] test this in a couple of services before merging
* [x] add documentation
* [x] further develop the response body logging API

### CODE REVIEW

File: src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package nl.knaw.dans.lib.logging.servlet.body
+
+import javax.servlet.http.HttpServletResponse
+import nl.knaw.dans.lib.logging.servlet.ResponseLogFormatter
+import org.scalatra.{ ActionResult, ScalatraBase }
+
+private[servlet] trait LogResponseBody extends ResponseLogFormatter {
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala:24]: Why not also request bodies? For example deposit status changes are supposed to be small. You might have to check the size as the actual content might be out of our control.

Or any custom content like for example the number of uploaded files.

Perhaps rather `formatCustomLogContent` than `body`.

File: src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  this: ScalatraBase =>
+  
+  def shouldLogResponseBody(response: HttpServletResponse): Boolean
+
+  override protected def formatResponseLog(actionResult: ActionResult): String = {
+    val body = if (shouldLogResponseBody(response)) s"; body=[${actionResult.body}]"
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala]: This does not allow to filter just some specific content of the body.

File: src/main/scala/nl/knaw/dans/lib/logging/servlet/package.scala [hunk 3]
```diff
@@ -184,4 +184,8 @@ package object servlet {
   }
 
   type MaskedLogFormatter = masked.MaskedLogFormatter
+
+  type LogResponseBody = body.LogResponseBody
+  type LogResponseBodyAlways = body.LogResponseBodyAlways
+  type LogResponseBodyOnError = body.LogResponseBodyOnError
```
Comment (jo-pol) [inline src/main/scala/nl/knaw/dans/lib/logging/servlet/package.scala:193]: Perhaps add (one of) these to the examples somewhere around line 53 and 77?

Or perhaps move the comment black to the test examples. Perhaps with a readme between the tests for things like the rational.

File: src/test/scala/nl/knaw/dans/lib/logging/servlet/AbstractServletLoggerSpec.scala [hunk 4]
```diff
@@ -74,12 +74,11 @@ class AbstractServletLoggerSpec extends FlatSpec with Matchers with EmbeddedJett
 
       val requestLine :: responseLine :: Nil = resultLines
 
-      requestLine should startWith(s"GET http://localhost:$port$path")
+      requestLine should startWith(s"request GET http://localhost:$port$path")
       requestLine should include(s"remote=$formattedRemote;")
 
-      responseLine should startWith(s"GET http://localhost:$port$path returned status=200; ")
+      responseLine should startWith(s"response GET http://localhost:$port$path returned status=200; ")
```
Comment (jo-pol) [inline src/test/scala/nl/knaw/dans/lib/logging/servlet/AbstractServletLoggerSpec.scala]: I thought you combined the headers, not drop them. So I expected something like
`responseLine should include(s"headers=[]")`

Same in the concrete ServletLoggerSpec.

Comment (rvanheest) [reply src/test/scala/nl/knaw/dans/lib/logging/servlet/AbstractServletLoggerSpec.scala]: In these concrete tests in `AbstractServletLoggerSpec`, there are two headers: `Date` and `Content-Type`. The former is (obviously) different for every run and the latter is tested for in the next line.

https://github.com/DANS-KNAW/dans-scala-lib/blob/5a49db9f5eedd86623871a9ba36881ccdb83203f/src/test/scala/nl/knaw/dans/lib/logging/servlet/AbstractServletLoggerSpec.scala#L81

The full response looks something like:

```
response GET http://localhost:65109/combinedLogger returned status=200; headers=[Date -> [Tue, 05 Mar 2019 08:20:23 GMT], Content-Type -> [text/plain;charset=utf-8]]
```

So, the headers are present, I did not remove them, even though it almost looks like it in

https://github.com/DANS-KNAW/dans-scala-lib/pull/17/files#diff-acef33b8123f7db626862068939d5ceaL82

To be sure I changed the test here as well:

```diff
-responseLine should startWith(s"response GET http://localhost:$port$path returned status=200; ")
+responseLine should startWith(s"response GET http://localhost:$port$path returned status=200; headers=[")
```

And similar in `ServletLoggerSpec`:

```diff
(mockedLogger.info(_: String)) expects where {
  s: String =>
-    (s startsWith s"response GET http://localhost:$port$testLoggerPath returned status=200") &&
+    (s startsWith s"response GET http://localhost:$port$testLoggerPath returned status=200; headers=[") &&
      (s.toLowerCase contains "content-type -> [text/plain;charset=utf-8]")
} once()
```

The full log line with merged headers is primarily tested in

https://github.com/DANS-KNAW/dans-scala-lib/pull/17/files#diff-222697fe1f39182d7abe71cfb9fbfc71R57

**Comment by rvanheest** [comment]
@jo-pol and I discussed the `LogResponseBody` feature yesterday. A first improvement we came up with is to not have a `shouldLogResponseBody` but a method like:

```scala
def formatResponseBody(actionResult: ActionResult): String
```

This solution is described in 12f40a0.

One problem with this is that it applies a certain response body formatting to all servlet endpoints declared in a `ScalatraServlet` file. While this is generally good enough for other formatting/logging such as masking, it might be that we require more fine-grained control over this when it comes to the response body.

In a future commit within this PR I'll try to develop that idea somewhat further.

File: src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala (hunk 1 above)
Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala:24]: @jo-pol and I decided not to log the request body. If you want to see stuff from the request body, add your own logs once you parsed the data.

File: src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala (hunk 2 above)
Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala]: @jo-pol and I moved away from `shouldLogResponseBody` and decided on `formatResponseBody`.

Comment (rvanheest) [reply src/main/scala/nl/knaw/dans/lib/logging/servlet/body/LogResponseBody.scala]: see also https://github.com/DANS-KNAW/dans-scala-lib/pull/17#issuecomment-470011496

File: src/test/scala/nl/knaw/dans/lib/logging/servlet/ResponseLogFormatterSpec.scala [hunk 5]
```diff
@@ -48,14 +48,45 @@ class ResponseLogFormatterSpec extends FlatSpec with Matchers with MockFactory w
...
+  }
+
+  it should "add no response body when using LogResponseBody with response without a body" in {
+    val testServlet: TestServlet = new TestServlet() with LogResponseBodyAlways {
+      // make formatResponseLog public again
+      override def formatResponseLog(actionResult: ActionResult): String = super.formatResponseLog(actionResult)
+    }
+    testServlet.formatResponseLog(NoContent()) shouldBe
+      "response GET http://does.not.exist.dans.knaw.nl returned status=204; headers=[Set-Cookie -> [scentry.auth.default.user=abc456.pq.xy], REMOTE_USER -> [somebody], Expires -> [Thu, 01 Jan 1970 00:00:00 GMT]]"
   }
```
Comment (rvanheest) [inline src/test/scala/nl/knaw/dans/lib/logging/servlet/ResponseLogFormatterSpec.scala:91]: add test(s) with `LogResponseBodyOnError`

**Pull request merged**